src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs

```diff
@@ -0,0 +1,395 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Diagnostics.Tracing;
+using System.Globalization;
+using System.Runtime.Versioning;
+using System.Text;
+
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// This EventSource is intended to let out-of-process tools (such as dotnet-counters) do
+    /// ad-hoc monitoring for the new Instrument APIs. This source only supports one listener
+    /// at a time. Each new listener will overwrite the configuration about which metrics
```

Comment:
Isn't that broken?  This is saying if I add my own EventListener and happen to enable the events from this event source, things break?

Reply:
it is certainly a limitation in the scenario, but it shouldn't render it useless. The main intended use case for this is ad-hoc collection tools and we'd tell people don't try to run multiple of these tools concurrently. If they do it anyways we can identify that happened and give an error message. For production/always-on monitoring scenarios they would be better off using a library like OpenTelemetry to export the data continuously. I also have some thoughts how we can eliminate this limitation in .NET 7.

Reply:
Rather than having multiple session overwrite the configuration in 6, could we keep track of a union of requested metrics?

All sessions would observe additions to the metric collection and increases in the frequency.

e.g.,

Session A enables Metric1;Metric2 @ 10s intervals.
Session B enables Metric1;Metric3 @  5s intervals.

then both sessions would see Metric1;Metric2;Metric3 @ 5s intervals.

Session B disables.

Session A continues to see Metric1;Metric2;Metric3 @ 5s intervals.

Then in 7 we could attempt to make changes unique to a session which would require more bookkeeping like your comment says.

Reply:
I think its possible, but it gets a little messy and I am not sure I have the time to work through the details. For example the requested intervals may not be multiples of one another, it requires the tools to re-filter, and it opens the door to having large numbers of metrics reported that won't stop reporting even after the tool disconnects.